File: lib/services/mock/mock_task_service.dart
import 'package:dartz/dartz.dart';
import '../../core/errors/failures.dart';
import '../../core/network/dio_client.dart';
import '../../models/task_model.dart';
import '../task_service.dart';
import 'mock_data.dart';

class MockTaskService extends TaskService {
  final List<TaskModel> _tasks = List.from(MockData.tasks);
  int _counter = MockData.tasks.length;

  MockTaskService(DioClient dioClient) : super(dioClient);

  @override
  Future<Either<Failure, List<TaskModel>>> getTasks({
    int page = 1,
    int pageSize = 20,
    String? statut,
  }) async {
    await Future.delayed(const Duration(milliseconds: 600));

    var filtered = _tasks.toList();
    if (statut != null) {
      filtered = filtered.where((t) => t.statut == statut).toList();
    }

    final start = (page - 1) * pageSize;
    if (start >= filtered.length) return const Right([]);

    final end = start + pageSize;
    return Right(filtered.sublist(start, end.clamp(0, filtered.length)));
  }

  @override
  Future<Either<Failure, TaskModel>> createTask(TaskModel task) async {
    await Future.delayed(const Duration(milliseconds: 800));

    _counter++;
    final newTask = TaskModel(
      id: 'tsk-${_counter.toString().padLeft(3, '0')}',
      numero: 'T-2026-${_counter.toString().padLeft(4, '0')}',
      codeClient: task.codeClient,
      nomClient: task.nomClient,
      adresse: task.adresse,
      description: task.description,
      datePrevue: task.datePrevue,
      priorite: task.priorite,
      statut: 'en_cours_traitement',
      commercialId: 'usr-001',
      commercialNom: 'Taha Mejdoub',
      createdAt: DateTime.now(),
    );
    _tasks.insert(0, newTask);
    return Right(newTask);
  }

  @override
  Future<Either<Failure, TaskModel>> updateTaskStatus(
    String taskId,
    String statut,
  ) async {
    await Future.delayed(const Duration(milliseconds: 500));

    final index = _tasks.indexWhere((t) => t.id == taskId);
    if (index == -1) {
      return const Left(ServerFailure(message: 'Tâche introuvable', statusCode: 404));
    }

    final old = _tasks[index];
    final updated = TaskModel(
      id: old.id,
      numero: old.numero,
      codeClient: old.codeClient,
      nomClient: old.nomClient,
      adresse: old.adresse,
      description: old.description,
      datePrevue: old.datePrevue,
      priorite: old.priorite,
      statut: statut,
      pieceJointeUrl: old.pieceJointeUrl,
      commercialId: old.commercialId,
      commercialNom: old.commercialNom,
      createdAt: old.createdAt,
    );
    _tasks[index] = updated;
    return Right(updated);
  }
}

File: olympia_admin/src/services/mock/mockTaskService.js
import { MOCK_TASKS } from '../../data/mockData.js';

// In-memory task store — mutated by create/update/delete
let _tasks = MOCK_TASKS.map((t) => ({ ...t }));

const mockTaskService = {
  async getTasks({ page = 1, pageSize = 20, statut } = {}) {
    await _delay(400);
    let list = statut ? _tasks.filter((t) => t.statut === statut) : [..._tasks];
    const total = list.length;
    const start = (page - 1) * pageSize;
    list = list.slice(start, start + pageSize);
    return { data: { data: list, total, page, pageSize }, error: null };
  },

  async getTaskById(id) {
    await _delay(200);
    const task = _tasks.find((t) => t.id === id);
    if (!task) return { data: null, error: { message: 'Tâche introuvable', code: 404 } };
    return { data: task, error: null };
  },

  async createTask(task) {
    await _delay(500);
    const newTask = {
      ...task,
      id: `tsk-${String(_tasks.length + 1).padStart(3, '0')}`,
      numero: `T-2026-${String(_tasks.length + 1).padStart(4, '0')}`,
      statut: 'en_cours_traitement',
      createdAt: new Date().toISOString().slice(0, 10),
    };
    _tasks = [newTask, ..._tasks];
    return { data: newTask, error: null };
  },

  async updateTaskStatus(id, statut) {
    await _delay(400);
    const idx = _tasks.findIndex((t) => t.id === id);
    if (idx === -1) return { data: null, error: { message: 'Tâche introuvable', code: 404 } };
    _tasks[idx] = { ..._tasks[idx], statut };
    return { data: _tasks[idx], error: null };
  },

  async updateTask(id, patch) {
    await _delay(400);
    const idx = _tasks.findIndex((t) => t.id === id);
    if (idx === -1) return { data: null, error: { message: 'Tâche introuvable', code: 404 } };
    _tasks[idx] = { ..._tasks[idx], ...patch };
    return { data: _tasks[idx], error: null };
  },

  async deleteTask(id) {
    await _delay(300);
    const idx = _tasks.findIndex((t) => t.id === id);
    if (idx === -1) return { data: null, error: { message: 'Tâche introuvable', code: 404 } };
    _tasks = _tasks.filter((t) => t.id !== id);
    return { data: null, error: null };
  },
};

function _delay(ms) { return new Promise((r) => setTimeout(r, ms)); }

export default mockTaskService;
